File: php-vs-python/fast_sort.php
<?PHP
  $num_arr = array();
  $fname = 'nums.dat';

  echo "start sorting!\r\n";
  $stime = microtime_float();

  echo "STEP 1: find_min_max--------\r\n";
  list($min_v, $max_v) = find_min_max( $fname );
  $rd = range( $min_v, $max_v-1, 1024 );
  $rd[] = $max_v;
  $len = count( $rd );

  foreach( $rd as $k => $v ) {
    if( $k>=($len-1) )
      break;

    echo "STEP 2: get_range--------\r\n";
    $num_arr = get_range( $v, $rd[$k+1], $fname );

    echo "STEP 3: fast_sort--------\r\n";
    fsort( 0, count($num_arr)-1 );
    show_arr( $num_arr );
  }

  $etime = microtime_float();

  $check_stime = microtime_float();
  $res = sort_check( $num_arr );
  $check_etime = microtime_float();

  if( $res==true ) {
    echo "\r\nUse ".($etime-$stime)."s to sort\r\n";
    echo "\r\nUse ".($check_etime-$check_stime)."s to check\r\n";
  }
  else
    echo "Sort failed!\r\n";

//==============================================================================
  function fsort( $sid, $eid ) {

    global $num_arr;

    if( $sid==$eid || $eid<$sid )
      return;

     $res_arr = classify( $sid, $eid );
     if( count($res_arr)<=0 )
      return;

     list( $l_sid, $l_eid, $r_sid, $r_eid ) = $res_arr;

     fsort( $l_sid, $l_eid );
     fsort( $r_sid, $r_eid );
  }

  function classify( $sid, $eid ) {

    global $num_arr;

    if( $sid==$eid || $sid>$eid )
      return [];

    $base = $num_arr[$eid];
    //echo "base: $base\r\n";
    $left_a = array();
    $right_a = array();
    for( $i=$sid; $i<$eid; $i++ ) {
        if( $num_arr[$i]<=$base )
          $left_a[] = $num_arr[$i];
        else
          $right_a[] = $num_arr[$i];
    }

    $mid_a = array_merge( $left_a, [$base], $right_a );
    array_splice( $num_arr, $sid, $eid-$sid+1, $mid_a );

    $l_sid = $sid;
    $l_eid = $l_sid + count($left_a) - 1;
    $r_sid = $l_eid + 2;
    $r_eid = $r_sid + count($right_a) - 1;

    return [ $l_sid, $l_eid, $r_sid, $r_eid ];
  }

  function show_arr( $arr ) {
    $i = 0;
    echo "The elements in array is:\r\n";
    foreach( $arr as $v ) {
      echo "$v ";
      $i++;
      if( $i==10 ) {
        echo "\r\n";
        $i = 0;
      }
    }

    echo "\r\n\r\n";
  }

  function sort_check( $arr ) {

    $eid = count( $arr ) - 1;
    $i = 0;

    foreach( $arr as $v1 ) {
      if( $i>=$eid )
        break;

      $c_arr = array_slice( $arr, $i+1 );

      foreach( $c_arr as $v2 ) {
        if( $v1>$v2 )
          return false;
      }
      $i++;
    }

    return true;
  }

  function microtime_float() {
    list($usec, $sec) = explode(" ", microtime());
    return ((float)$usec + (float)$sec);
  }

function find_min_max( $fname ) {

  $fh = fopen( $fname, "rb" );
  $res = [ 0, 0 ];

  while( 1 ) {
    $data_str = '';
    $data_str = fread( $fh, 2*pow(2,20) );
    if( empty($data_str) )
      break;

      $slen = strlen( $data_str );
      $i = 0;
      while( 1 ) {
        if( ($i*4)>=$slen )
          break;
        $mid = unpack( "l", substr($data_str,$i*4,4) );
        if( $mid[1]<$res[0] ) {
          $res[0] = $mid[1];
          $i++;
          continue;
        }

        if( $mid[1]>$res[1] )
          $res[1] = $mid[1];

        $i++;
      }
  }

  fclose( $fh );
  //var_dump( $res );
  return $res;
}

function get_range( $min_v, $max_v, $fname ) {

  $fh = fopen( $fname, "rb" );
  $res = [];

  while( 1 ) {
    $data_str = '';
    $data_str = fread( $fh, 2*pow(2,20) );
    if( empty($data_str) )
      break;

      $slen = strlen( $data_str );
      $i = 0;
      while( 1 ) {
        if( ($i*4)>=$slen )
          break;
        $mid = unpack( "l", substr($data_str,$i*4,4) );
        if( $mid[1]>=$min_v && $mid[1]<$max_v )
          $res[] = $mid[1];

        $i++;
      }
  }

  fclose( $fh );
  return $res;
}

?>

File: php-vs-python/fast_sort.py
import random
import time
import timeit
import struct
import sys

def show_arr( arr ):
    num = 0
    for i in arr:
        print i,
        num += 1
        if num==10:
            print
            num = 0

def classify( sid, eid ):
    global arr_num

    num = len( arr_num )

    if sid>=eid:
        return 0,0,0,0

    l_arr = []
    r_arr = []
    base = arr_num[eid]

    for i in arr_num[sid:eid]:
        if i<=base:
            l_arr.append( i )
        else:
            r_arr.append( i )

    l_sid = sid
    l_eid = l_sid + len(l_arr) - 1
    r_sid = l_eid + 2
    r_eid = r_sid + len(r_arr) - 1

    arr_1 = []
    if (sid-1)>=0:
        arr_1 = arr_num[0:sid]

    arr_2 = []
    if (eid+1)<=(num-1):
        arr_2 = arr_num[eid+1:]

    arr_num = arr_1 + l_arr + [base] + r_arr + arr_2

    return l_sid,l_eid,r_sid,r_eid

def f_sort( sid, eid ):
    if sid>=eid:
        return

    l_sid,l_eid,r_sid,r_eid = classify( sid, eid )
    f_sort( l_sid, l_eid )
    f_sort( r_sid, r_eid )

def find_min_max( fname ):
    min_v = 0
    max_v = 0
    data_str = ''

    fp = open( fname, 'rb' )
    while 1:
        data_str = fp.read( 1024*8 )
        slen = len( data_str )
        if slen<=0:
            break

        for i in range(0, slen, 4):
            mid = struct.unpack( "i", data_str[i:i+4] )[0]
            if mid<min_v:
                min_v = mid
                continue;

            if mid>max_v:
                max_v = mid

        data_str = ''

    fp.close()
    return min_v, max_v

def get_range( min_v, max_v, fname ):
    res = [];
    data_str = ''

    fp = open( fname, 'rb' )
    while 1:
        data_str = fp.read( 1024*8 )
        slen = len( data_str )
        if slen<=0:
            break

        for i in range(0, slen, 4):
            mid = struct.unpack( "i", data_str[i:i+4] )[0]
            if min_v <= mid < max_v:
                res += [ mid ]
                continue;

        data_str = ''

    fp.close()
    return res

#-------------------------------------------------------------------------------
fname = 'nums.dat'

print "Start sorting!"
start = timeit.default_timer()

min_v, max_v = find_min_max( fname )
rd = range(min_v,max_v,1024*10) + [ max_v ]
for i, v in enumerate( rd[0:-1] ):
    arr_num = get_range( v, rd[i+1], fname )
    f_sort( 0, len(arr_num)-1 )
    print arr_num

elapsed = (timeit.default_timer() - start)
print "Use %f s" % elapsed
